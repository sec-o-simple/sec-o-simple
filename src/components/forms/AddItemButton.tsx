import { faAdd } from '@fortawesome/free-solid-svg-icons'
import {
  FontAwesomeIcon,
  FontAwesomeIconProps,
} from '@fortawesome/react-fontawesome'
import { Button, ButtonProps } from '@heroui/button'
import { twMerge } from 'tailwind-merge'

export type AddItemButtonProps = ButtonProps & {
  label?: string
  icon?: FontAwesomeIconProps['icon']
}

export default function AddItemButton(props: AddItemButtonProps) {
  const { label = 'Add New Item', icon = faAdd, ...buttonProps } = props
  return (
    <Button
      {...buttonProps}
      variant="bordered"
      className={twMerge(
        'border-dashed text-neutral-foreground border-gray',
        buttonProps.className ?? '',
      )}
    >
      <FontAwesomeIcon icon={icon} />
      {label}
    </Button>
  )
}
