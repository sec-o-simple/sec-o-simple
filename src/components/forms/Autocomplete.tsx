import { useDebounceInput } from '@/utils/useDebounceInput'
import { useFieldValidation } from '@/utils/useFieldValidation'
import {
  Autocomplete as HeroUIAutocomplete,
  AutocompleteProps,
} from '@heroui/react'

export function Autocomplete(
  props: AutocompleteProps & { csafPath?: string; isTouched?: boolean },
) {
  const {
    placeholder,
    labelPlacement,
    variant,
    csafPath,
    isTouched,
    onBlur,
    onChange,
    onValueChange,
    value: propValue,
    ...rest
  } = props
  const validation = useFieldValidation(csafPath)

  const { handleChange } = useDebounceInput({
    onChange: (e) => {
      onValueChange?.(e.target.value)
      onChange?.(e)
    },
    onBlur,
    value: propValue as string,
  })

  return (
    <HeroUIAutocomplete
      labelPlacement={labelPlacement ?? 'outside'}
      placeholder={placeholder ?? ' '}
      variant={variant ?? 'bordered'}
      inputProps={{
        classNames: {
          inputWrapper: 'border-1 shadow-none',
        },
      }}
      onChange={handleChange}
      errorMessage={validation.errorMessages.map((m) => m.message).join(', ')}
      isInvalid={validation.hasErrors && (isTouched || validation.isTouched)}
      {...rest}
    />
  )
}
