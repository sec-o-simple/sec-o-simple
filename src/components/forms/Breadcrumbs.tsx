import { Breadcrumbs as MuiBreadcrumbs } from '@heroui/react'

export default function Breadcrumbs({
  children,
}: Readonly<{
  children: React.ReactNode
}>) {
  return (
    <MuiBreadcrumbs
      maxItems={10}
      radius="lg"
      variant="solid"
      classNames={{
        list: 'bg-white border-1 border-default-200',
      }}
    >
      {children}
    </MuiBreadcrumbs>
  )
}
