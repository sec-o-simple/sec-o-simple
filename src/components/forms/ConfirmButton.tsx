import { Button, ButtonProps } from '@heroui/button'
import {
  Modal,
  ModalBody,
  ModalContent,
  ModalFooter,
  ModalHeader,
  useDisclosure,
} from '@heroui/react'
import { useTranslation } from 'react-i18next'

type ConfirmButtonProps = {
  confirmText: string
  confirmTitle: string
  onConfirm: () => void
  skipConfirm?: boolean
} & ButtonProps

export default function ConfirmButton({
  confirmText,
  confirmTitle,
  onConfirm,
  skipConfirm = false,
  ...props
}: ConfirmButtonProps) {
  const { t } = useTranslation()
  const { isOpen, onOpen, onOpenChange } = useDisclosure()

  return (
    <>
      <Button onPress={onOpen} fullWidth {...props}>
        {props?.children || t('common.confirm')}
      </Button>
      <Modal isOpen={isOpen} onOpenChange={onOpenChange} size="xl">
        <ModalContent>
          {(onClose) => (
            <>
              <ModalHeader className="flex flex-col gap-1">
                {confirmTitle || t('common.confirm')}
              </ModalHeader>
              <ModalBody className="gap-4">
                <p>{confirmText}</p>
              </ModalBody>
              <ModalFooter>
                <Button variant="light" onPress={onClose}>
                  {t('common.cancel')}
                </Button>
                <Button
                  color="primary"
                  onPress={() => {
                    onConfirm()
                    onClose()
                  }}
                >
                  {t('common.confirm')}
                </Button>
              </ModalFooter>
            </>
          )}
        </ModalContent>
      </Modal>
    </>
  )
}
