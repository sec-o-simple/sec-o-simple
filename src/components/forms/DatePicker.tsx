import { useFieldValidation } from '@/utils/validation/useFieldValidation'
import {
  DatePicker as HeroUIDatePicker,
  DatePickerProps as HeroUIDatePickerProps,
} from '@heroui/date-picker'
import { getLocalTimeZone, parseAbsolute } from '@internationalized/date'

export type DatePickerProps = Omit<
  HeroUIDatePickerProps,
  'value' | 'onChange'
> & {
  csafPath?: string
  isTouched?: boolean
  value?: string
  onChange?: (date: string) => void
}

export default function DatePicker({
  labelPlacement,
  variant,
  csafPath,
  isTouched,
  onBlur,
  onChange,
  value,
  ...pickerProps
}: DatePickerProps) {
  const validation = useFieldValidation(csafPath)

  return (
    <HeroUIDatePicker
      variant={variant ?? 'bordered'}
      labelPlacement={labelPlacement ?? 'outside'}
      classNames={{
        inputWrapper: 'border-1 shadow-none',
      }}
      errorMessage={validation.errorMessages.map((m) => m.message).join(', ')}
      isInvalid={
        validation.hasErrors &&
        (isTouched || validation.isTouched || !!value?.length)
      }
      value={value ? parseAbsolute(value, getLocalTimeZone()) : undefined}
      onBlur={(e) => {
        if (csafPath) {
          validation.markFieldAsTouched(csafPath)
        }
        onBlur?.(e)
      }}
      onChange={(dateValue) =>
        onChange?.(dateValue?.toDate(getLocalTimeZone()).toISOString() ?? '')
      }
      {...pickerProps}
    />
  )
}
