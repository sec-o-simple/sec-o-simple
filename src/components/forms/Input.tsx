import { useDebounceInput } from '@/utils/useDebounceInput'
import { useFieldValidation } from '@/utils/validation/useFieldValidation'
import {
  Input as HeroUIInput,
  Textarea as HeroUITextarea,
  InputProps,
  TextAreaProps,
} from '@heroui/input'

export function Input(
  props: InputProps & { csafPath?: string; isTouched?: boolean },
) {
  const {
    placeholder,
    labelPlacement,
    variant,
    csafPath,
    isTouched,
    onBlur,
    onChange,
    onValueChange,
    isInvalid,
    value: propValue,
    ...rest
  } = props
  const validation = useFieldValidation(csafPath)

  const {
    value: debouncedValue,
    isDebouncing,
    handleBlur,
    handleChange,
  } = useDebounceInput({
    onChange: (e) => {
      onValueChange?.(e.target.value)
      onChange?.(e)
    },
    onBlur,
    value: propValue,
  })

  return (
    <HeroUIInput
      variant={variant ?? 'bordered'}
      labelPlacement={labelPlacement ?? 'outside'}
      placeholder={placeholder ?? ' '}
      classNames={{
        inputWrapper: 'border-1 shadow-none',
      }}
      value={debouncedValue}
      errorMessage={validation.errorMessages.map((m) => m.message).join(', ')}
      isInvalid={
        isInvalid ||
        (!isDebouncing &&
          validation.hasErrors &&
          (isTouched || validation.isTouched || !!propValue?.length))
      }
      onBlur={(e) => {
        if (csafPath) {
          validation.markFieldAsTouched(csafPath)
        }
        handleBlur(e)
      }}
      onChange={handleChange}
      {...rest}
    />
  )
}

export function Textarea(
  props: TextAreaProps & { csafPath?: string; isTouched?: boolean },
) {
  const {
    placeholder,
    labelPlacement,
    variant,
    csafPath,
    isTouched,
    onBlur,
    onChange,
    onValueChange,
    value: propValue,
    ...rest
  } = props
  const validation = useFieldValidation(csafPath)
  const {
    value: debouncedValue,
    isDebouncing,
    handleBlur,
    handleChange,
  } = useDebounceInput({
    onChange: (e) => {
      onValueChange?.(e.target.value)
      onChange?.(e)
    },
    onBlur,
    value: propValue,
  })

  return (
    <HeroUITextarea
      variant={variant ?? 'bordered'}
      labelPlacement={labelPlacement ?? 'outside'}
      placeholder={placeholder ?? ' '}
      classNames={{
        inputWrapper: 'border-1 shadow-none',
      }}
      value={debouncedValue}
      errorMessage={validation.errorMessages.map((m) => m.message).join(', ')}
      isInvalid={
        !isDebouncing &&
        validation.hasErrors &&
        (validation.isTouched || isTouched)
      }
      onBlur={(e) => {
        if (csafPath) {
          validation.markFieldAsTouched(csafPath)
        }
        handleBlur(e)
      }}
      onChange={handleChange}
      {...rest}
    />
  )
}
