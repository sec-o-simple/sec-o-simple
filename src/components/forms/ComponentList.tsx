import { faTrash } from '@fortawesome/free-solid-svg-icons'
import { Accordion, AccordionItem } from '@heroui/accordion'
import { HTMLProps, ReactNode, useState } from 'react'
import IconButton from './IconButton'
import { Selection } from '@heroui/react'
import { ListState } from '@/utils/useListState'
import {
  DynamicObjectValueKey,
  getDynamicObjectValue,
} from '@/utils/dynamicObjectValue'
import AddItemButton from './AddItemButton'
import { checkReadOnly } from '@/utils/template'
import { twMerge } from 'tailwind-merge'

export type ComponentListProps<T> = {
  listState: ListState<T>
  title: DynamicObjectValueKey<T>
  /** Generator function for a ReactNode that will be shown when an item is expanded */
  content: (item: T) => ReactNode
  onChange?: (updatedItems: T[]) => void
  onDelete?: (item: T) => void
  startContent?: (item: T) => ReactNode
  endContent?: (item: T) => ReactNode
  titleProps?: HTMLProps<HTMLDivElement>
}

export default function ComponentList<T extends object>({
  listState,
  title,
  content,
  onDelete,
  startContent,
  endContent,
  titleProps,
}: ComponentListProps<T>) {
  const [expandedKeys, setExpandedKeys] = useState<Selection>(new Set([]))

  return (
    <div className="flex flex-col gap-2">
      <Accordion
        selectedKeys={expandedKeys}
        onSelectionChange={setExpandedKeys}
        variant="splitted"
        selectionMode="multiple"
        itemClasses={{
          trigger: 'flex justify-between py-0',
          titleWrapper: 'hidden',
          startContent: 'grow',
          indicator: 'text-neutral-foreground',
          base: 'border border-gray shadow-none px-4 py-2',
          content: 'py-2',
        }}
        className="px-0"
      >
        {listState.data.map((item) => (
          <AccordionItem
            key={listState.getId(item)}
            startContent={
              <div className="flex items-center justify-between">
                <div className="flex items-center gap-4">
                  {startContent?.(item)}
                  <div
                    {...titleProps}
                    className={twMerge(
                      'max-w-xl overflow-hidden text-ellipsis text-nowrap',
                      titleProps?.className,
                    )}
                  >
                    {getDynamicObjectValue(item, title) || (
                      <span>Untitled</span>
                    )}
                  </div>
                  {endContent?.(item)}
                </div>
                <IconButton
                  icon={faTrash}
                  onPress={() =>
                    onDelete
                      ? onDelete?.(item)
                      : listState.removeDataEntry(item)
                  }
                  isDisabled={checkReadOnly(item)}
                />
              </div>
            }
          >
            {content(item)}
          </AccordionItem>
        ))}
      </Accordion>
      <AddItemButton
        onPress={() => {
          const key = listState.addDataEntry()
          // expand new item
          if (key) {
            setExpandedKeys(new Set([...expandedKeys, key]))
          }
        }}
      />
    </div>
  )
}
