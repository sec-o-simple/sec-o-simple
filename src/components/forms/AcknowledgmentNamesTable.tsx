import { Input } from '@/components/forms/Input'
import {
  getDefaultDocumentAcknowledgmentName,
  TAcknowledgment,
  TAcknowledgmentName,
} from '@/routes/document-information/types/tDocumentAcknowledgments'
import { useListState } from '@/utils/useListState'
import { faAdd, faTrash } from '@fortawesome/free-solid-svg-icons'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { Button } from '@heroui/button'
import {
  Table,
  TableBody,
  TableCell,
  TableColumn,
  TableHeader,
  TableRow,
} from '@heroui/table'
import { useEffect } from 'react'
import { useTranslation } from 'react-i18next'
import IconButton from './IconButton'

export default function AcknowledgmentNamesTable({
  acknowledgment,
  acknowledgmentIndex,
  onChange,
}: {
  acknowledgment: TAcknowledgment
  acknowledgmentIndex: number
  onChange: (acknowledgment: TAcknowledgment) => void
}) {
  const { t } = useTranslation()
  const acknowledgmentNames = useListState<TAcknowledgmentName>({
    initialData: acknowledgment.names,
    generator: getDefaultDocumentAcknowledgmentName,
  })

  useEffect(
    () =>
      onChange({
        ...acknowledgment,
        names: acknowledgmentNames.data,
      }),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [acknowledgmentNames.data],
  )

  const handleAddRevision = () => acknowledgmentNames.addDataEntry()
  const handleRevisionChange = (updatedName: TAcknowledgmentName) =>
    acknowledgmentNames.updateDataEntry(updatedName)
  const handleDeleteName = (name: TAcknowledgmentName) =>
    acknowledgmentNames.removeDataEntry(name)

  return (
    <div className="mt-4">
      <div className="mb-4 flex items-center justify-between">
        <h2 className="text-lg font-semibold">
          {t('document.acknowledgments.names')}
        </h2>
        <Button variant="light" color="primary" onPress={handleAddRevision}>
          <FontAwesomeIcon icon={faAdd} />
          {t('common.add', {
            label: t('document.acknowledgments.name'),
          })}
        </Button>
      </div>

      <Table>
        <TableHeader>
          <TableColumn width="20%">
            {t('document.acknowledgments.name')}
          </TableColumn>
          <TableColumn width="10%">{t('common.actions')}</TableColumn>
        </TableHeader>
        <TableBody emptyContent={t('document.acknowledgments.empty')}>
          {acknowledgmentNames.data?.map((name, index) => (
            <TableRow key={name.name}>
              <TableCell>
                <Input
                  value={name.name}
                  csafPath={`/document/acknowledgments/${acknowledgmentIndex}/names/${index}`}
                  onValueChange={(value) =>
                    handleRevisionChange({
                      ...acknowledgmentNames.data[index],
                      name: value,
                    })
                  }
                  placeholder={t('common.placeholder', {
                    label: t('document.acknowledgments.name'),
                  })}
                />
              </TableCell>
              <TableCell>
                <IconButton
                  icon={faTrash}
                  onPress={() => handleDeleteName(name)}
                />
              </TableCell>
            </TableRow>
          ))}
        </TableBody>
      </Table>
    </div>
  )
}
