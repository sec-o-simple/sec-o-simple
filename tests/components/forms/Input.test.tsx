import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { describe, it, expect, vi, beforeEach } from 'vitest'
import { Input, Textarea } from '../../../src/components/forms/Input'

describe('Input', () => {
  beforeEach(() => {
    vi.clearAllMocks()
  })

  it('should render with default props', () => {
    const { container } = render(<Input />)
    
    const input = screen.getByTestId('hero-input')
    expect(input).toBeInTheDocument()
    expect(input).toHaveAttribute('data-variant', 'bordered')
    expect(input).toHaveAttribute('data-labelplacement', 'outside')
    expect(input).toHaveAttribute('placeholder', ' ')
    expect(container.firstChild).toMatchSnapshot()
  })

  it('should accept custom props', () => {
    render(
      <Input
        label="Custom Label"
        placeholder="Custom placeholder"
        variant="underlined"
        labelPlacement="inside"
      />
    )
    
    const input = screen.getByTestId('hero-input')
    expect(input).toHaveAttribute('label', 'Custom Label')
    expect(input).toHaveAttribute('placeholder', 'Custom placeholder')
    expect(input).toHaveAttribute('data-variant', 'underlined')
    expect(input).toHaveAttribute('data-labelplacement', 'inside')
  })

  it('should render with controlled value', () => {
    render(<Input value="controlled value" />)
    
    const input = screen.getByTestId('hero-input')
    expect(input).toHaveAttribute('value', 'controlled value')
  })

  it('should apply custom classNames', () => {
    render(<Input />)
    
    const input = screen.getByTestId('hero-input')
    expect(input).toHaveAttribute('data-classnames', JSON.stringify({
      inputWrapper: 'border-1 shadow-none',
    }))
  })

  it('should handle invalid state correctly', () => {
    const { container } = render(<Input isInvalid />)
    
    const input = screen.getByTestId('hero-input')
    expect(input).toHaveAttribute('data-isinvalid', 'true')
    expect(container.firstChild).toMatchSnapshot()
  })

  it('should handle error message prop', () => {
    render(<Input errorMessage="Custom error" />)
    
    const input = screen.getByTestId('hero-input')
    expect(input).toHaveAttribute('data-errormessage', 'Custom error')
  })

  it('should pass through additional props', () => {
    render(<Input disabled id="test-input" />)
    
    const input = screen.getByTestId('hero-input')
    expect(input).toHaveAttribute('disabled')
    expect(input).toHaveAttribute('id', 'test-input')
  })

  it('should render with csafPath prop', () => {
    render(<Input csafPath="test.field" />)
    
    const input = screen.getByTestId('hero-input')
    expect(input).toBeInTheDocument()
    // The csafPath is used internally for validation but doesn't affect the DOM
  })
})

describe('Textarea', () => {
  beforeEach(() => {
    vi.clearAllMocks()
  })

  it('should render with default props', () => {
    const { container } = render(<Textarea />)
    
    const textarea = screen.getByTestId('hero-textarea')
    expect(textarea).toBeInTheDocument()
    expect(textarea).toHaveAttribute('data-variant', 'bordered')
    expect(textarea).toHaveAttribute('data-labelplacement', 'outside')
    expect(textarea).toHaveAttribute('placeholder', ' ')
    expect(container.firstChild).toMatchSnapshot()
  })

  it('should accept custom props', () => {
    render(
      <Textarea
        label="Custom Label"
        placeholder="Custom placeholder"
        variant="underlined"
        labelPlacement="inside"
      />
    )
    
    const textarea = screen.getByTestId('hero-textarea')
    expect(textarea).toHaveAttribute('label', 'Custom Label')
    expect(textarea).toHaveAttribute('placeholder', 'Custom placeholder')
    expect(textarea).toHaveAttribute('data-variant', 'underlined')
    expect(textarea).toHaveAttribute('data-labelplacement', 'inside')
  })

  it('should render with controlled value', () => {
    render(<Textarea value="controlled textarea value" />)
    
    const textarea = screen.getByTestId('hero-textarea')
    expect(textarea).toBeInTheDocument()
    // The value prop is handled internally by the component
  })

  it('should handle error message prop', () => {
    render(<Textarea errorMessage="Textarea error" />)
    
    const textarea = screen.getByTestId('hero-textarea')
    expect(textarea).toHaveAttribute('data-errormessage', 'Textarea error')
  })

  it('should apply custom classNames', () => {
    render(<Textarea />)
    
    const textarea = screen.getByTestId('hero-textarea')
    expect(textarea).toHaveAttribute('data-classnames', JSON.stringify({
      inputWrapper: 'border-1 shadow-none',
    }))
  })

  it('should pass through additional props', () => {
    render(<Textarea disabled id="test-textarea" />)
    
    const textarea = screen.getByTestId('hero-textarea')
    expect(textarea).toHaveAttribute('disabled')
    expect(textarea).toHaveAttribute('id', 'test-textarea')
  })

  it('should render with csafPath prop', () => {
    render(<Textarea csafPath="test.textarea" />)
    
    const textarea = screen.getByTestId('hero-textarea')
    expect(textarea).toBeInTheDocument()
    // The csafPath is used internally for validation but doesn't affect the DOM
  })

  it('should handle isTouched prop', () => {
    render(<Textarea isTouched />)
    
    const textarea = screen.getByTestId('hero-textarea')
    expect(textarea).toBeInTheDocument()
    // The isTouched prop affects validation but doesn't directly change DOM attributes
  })
})
