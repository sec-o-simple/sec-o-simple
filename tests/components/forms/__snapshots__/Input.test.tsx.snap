// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Input > should handle invalid state correctly 1`] = `
<input
  data-classnames="{"inputWrapper":"border-1 shadow-none"}"
  data-errormessage=""
  data-isinvalid="true"
  data-labelplacement="outside"
  data-testid="hero-input"
  data-variant="bordered"
  placeholder=" "
  value=""
/>
`;

exports[`Input > should render with default props 1`] = `
<input
  data-classnames="{"inputWrapper":"border-1 shadow-none"}"
  data-errormessage=""
  data-isinvalid="false"
  data-labelplacement="outside"
  data-testid="hero-input"
  data-variant="bordered"
  placeholder=" "
  value=""
/>
`;

exports[`Textarea > should render with default props 1`] = `
<textarea
  data-classnames="{"inputWrapper":"border-1 shadow-none"}"
  data-errormessage=""
  data-isinvalid="false"
  data-labelplacement="outside"
  data-testid="hero-textarea"
  data-variant="bordered"
  placeholder=" "
/>
`;
